feat: added customizable refit preset support

zdev4 opened the pull request:
How to use:

It is now possible to create a global override of the baked in refit preset. To do that simply have a preset named either "Female-Refit" for female actors or "Male-Refit" for male actors. If not having such a preset in the preset list OBody will fall back to the baked in one.

It is also possible to create refit presets for each normal presets. Say you have a preset called "MyPreset". Now if you have another preset called "MyPreset-Refit" each character that has "MyPreset" as selected body will use "MyPreset-Refit" while clothed.

It is very important that refit presets must only contain the differences from the normal base preset. Say that "MyPreset" has value 10 for the Pushup slider. If your desired value for the Pushup slider is 50 then "MyPreset-Refit" must contain value 40 (50-10) for the Pushup slider.

Refit presets (having a suffix "-Refit" in their names) are blacklisted by default meaning they won't be included in random body distribution or appear in the OBody preset selection menu. They are solely used for the refit feature.

Commit:
feat: added customizable refit preset support

How to use:

It is now possible to create a global override of the baked in refit
preset. To do that simply have a preset named either "Female-Refit" for
female actors or "Male-Refit" for male actors. If not having such a
preset in the preset list OBody will fall back to the baked in one.

It is also possible to create refit presets for each normal presets.
Say you have a preset called "MyPreset". Now if you have another preset
called "MyPreset-Refit" each character that has "MyPreset" as selected
body will use "MyPreset-Refit" while clothed.

It is very important that refit presets must only contain the
differences from the normal base preset. Say that "MyPreset" has
value 10 for the Pushup slider. If your desired value for the Pushup
slider is 50 then "MyPreset-Refit" must contain value 40 (50-10) for
the Pushup slider.

Refit presets (having a suffix "-Refit" in their names) are blacklisted
by default meaning they won't be included in random body distribution
or appear in the OBody preset selection menu. They are solely used for
the refit feature.

Aietos commented:
Sorry, only had the chance to take a look at this today. Also because Github didn't notify me of a new pull request :(

This seems like a nice simple solution for a customisable ORefit!

Before I merge this I have a question, different outfits may have different refit needs even if the preset is the same. Do you think this solution would be easily expandable to also have a per-outfit refit setting, or this would require a more complex implementation?

I'm fine with merging this as is, just wondering how this could be expanded and trying to get new ideas for future improvements to this.

zdev4 commented:
Well, it would probably be doable with something like this in the JSON:

`"refitOutfitPresetsFemale": {
	"Steel Top I": "SomePreset1-Refit",
	"Leather Top I": "SomePreset2-Refit"
},
"refitOutfitPresetsMale": {
	"Steel Top I": "SomePreset3-Refit",
	"Leather Top I": "SomePreset4-Refit"
},`

or a different approach:

`"refitOutfitPresetsFemale": {
	"SomePreset1-Refit": [ "Steel Top I", "...", "..." ],
	"SomePreset2-Refit": [ "Leather Top I", "...", "..." ]
},
"refitOutfitPresetsMale": {
	"SomePreset3-Refit": [ "Steel Top I", "...", "..." ],
	"SomePreset4-Refit": [ "Leather Top I", "...", "..." ]
},`

so this would take priority over everything else when wearing that outfit.

The problem with this is the actor could wear multiple outfit pieces resolving to different presets. So only one should be selected from all the equipped outfit pieces, potentially the one in slot 32.

I haven't yet thoroughly checked how outfits and json keys are handled in the code but I could take a look if this solution is acceptable.

Commit:
feat: added a per outfit refit preset option

How to use:

This feature adds two new keys to OBody_presetDistributionConfig.json:

refitOutfitPresetsFemale can be used to specify female refit presets for
each outfit.

refitOutfitPresetsMale can be used to specify male refit presets for
each outfit.

For each key an object must be specified where keys are the names of
the outfits and values are the refit preset names.

example:

{
 "refitOutfitPresetsFemale": {
  "Outfit Name 1": "FemalePreset1-Refit",
  "Outfit Name 2": "FemalePreset1-Refit",
  "Outfit Name 3": "FemalePreset2-Refit"
 },
 "refitOutfitPresetsMale": {
  "Outfit Name 1": "MalePreset1-Refit",
  "Outfit Name 2": "MalePreset1-Refit",
  "Outfit Name 3": "MalePreset2-Refit"
 }
}

Per outfit refit presets take priority over all other refit preset
options.

zdev4 commented:
I've pushed a second commit containing the per-outfit preset options. I went with the first JSON format because that way it's faster to find the requested preset and it prevents mistakes by not letting to specify an outfit name under two different presets.

This solution only checks for outfits in the body and chest slots, but can be extended with other slots if needed later.

Aietos commented:
Changes look good, approved! Will be merged when the pipeline passes.

Thanks for implementing this feature, I will make a release on Nexus after I've tested it and also after writing the guide on how to use it :)
